onnxruntime/core/providers/cpu/tensor/upsample_aa.h

```diff
@@ -0,0 +1,682 @@
+          if (output_height == input_height) {
+            memcpy(Ydata + (n * num_channels) * (output_height * output_width), Xdata,
+                   sizeof(T) * narrow<size_t>(output_height * output_width));
+            return;
+          }
+          for (size_t y = 0; y < narrow<size_t>(output_height); ++y) {
+            const auto* weight_coeff =
+                reinterpret_cast<const ACtype*>(p.dim_y.weight_coefficients.get()) +
+                p.dim_y.window_size * y;
+            int64_t ymin = p.dim_y.bound[y * 2];
+            int64_t ymax = p.dim_y.bound[y * 2 + 1];
+
+            for (size_t x = 0; x < narrow<size_t>(output_width); ++x) {
+              const int64_t output_offset = output_width * y + x;
+              auto* Ydata_offset = Ydata + output_offset;
+
+              if (use_extrapolation &&
+                  ((p.dim_y.original[y] < 0 || p.dim_y.original[y] > static_cast<float>(input_height - 1)) ||
+                   (p.dim_x.original[x] < 0 || p.dim_x.original[x] > static_cast<float>(input_width - 1)))) {
+                *Ydata_offset = static_cast<T>(extrapolation_value);
+                continue;
+              }
+
+              ACtype output = is_8bit_v<T> ? ConstValue::mag_factor : 0;
+              auto* weight_coeff_start = weight_coeff;
+
+              const auto* Xdata_offset = Xdata + ymin * output_width + x;
+              for (auto idx = ymin; idx < ymax; ++idx) {
+                output += *Xdata_offset * (*weight_coeff_start++);
+                Xdata_offset += output_width;
+              }
+              if constexpr (is_8bit_v<T>) {
+                *Ydata_offset = static_cast<T>(clip8_lookups[output >> 22]);
+              } else if constexpr (std::is_same<T, int32_t>::value) {
+                *Ydata_offset = FilterParamsAA::round_up(output);
+              } else {  // float double
+                *Ydata_offset = static_cast<T>(output);
+              }
+            }
+          }
+        });
+  }
+}
+
+template <typename T>
+void UpsampleBilinearAA(const int64_t batch_size,
+                        const int64_t num_channels,
+                        const int64_t input_height,
+                        const int64_t input_width,
+                        const int64_t output_height,
+                        const int64_t output_width,
+                        const float height_scale,
+                        const float width_scale,
+                        const std::vector<float>& roi,
+                        const bool use_extrapolation,
+                        const float extrapolation_value,
+                        bool exclude_outside,
+                        const Tensor* const X,
+                        T* const YdataBase,
+                        AllocatorPtr& alloc,
+                        const GetOriginalCoordinateFunc& get_original_coordinate,
+                        concurrency::ThreadPool* tp) {
+  const auto* XdataBase = X->Data<T>();
```

Comment:
nit: We don't seem to do anything to change this value, so wouldn't directly passing through the Tensor be simpler?